content/en/blog/_posts/2024-11-11-dynamic-kubernetes-api-server-for-cozystack/index.md

```diff
@@ -0,0 +1,454 @@
+
+So, in our platform, everything is configured as `HelmRelease` resources. However, we ran into
+two problems: limitations of the RBAC model and the need for a public API. Let's delve into these
+
+### Limitations of the RBAC model
+
+The standard RBAC system in Kubernetes doesn't allow you to restrict access to a list of resources
+of the same kind based on labels or specific fields in the spec. When creating a role, you can limit
+access across the resources in the same kind only by specifying specific resource names in `resourceNames`.
+For methods like GET, PATCH, or UPDATE it will work. However, filtering by `resourceNames` in the LIST
+method doesn't work like that. Thus you can limit listing certain resources by kind but not by name.
+
+So, we decided to introduce new resource types based on the names of the Helm charts they use and
+generate the list of available kinds dynamically at runtime in our extension api-server.
+This way, we can reuse Kubernetes standard RBAC model to manage access to specific resource types.
+
+### Need for a public API
+
+Since our platform provides capabilities for deploying various managed services, we want to organize
+public access to the platform's API. However, we can't allow users to interact directly with resources
+like `HelmRelease` because that would let them specify arbitrary names and parameters for Helm charts to
+deploy, potentially compromising our system.
+
+We wanted to give users the ability to deploy a specific service simply by creating the resource with corresponding
+kind in Kubernetes. The type of this resource should be named the same as the chart from
+which it's deployed. Here are some examples:
+
+- `Kind: Kubernetes` → `chart: kubernetes`
+- `Kind: Postgres` → `chart: postgres`
+- `Kind: Redis` → `chart: redis`
+- `Kind: VirtualMachine` → `chart: virtual-machine`
+
+Moreover, we don't want to have to add a new type to codegen and recompile our extension api-server
+every time we add a new chart for it to start being served.
+The update should be done dynamicly or provided via a `ConfigMap` by the administrator.
+
+### Two-Way conversion
+
+Currently, we already have integrations and a dashboard that continue to use `HelmRelease` resources.
+At this stage, we didn't want to lose the ability to support this API. Considering that we're simply
+translating one resource into another, support is maintained and it works both ways.
+If you create a `HelmRelease`, you'll get a custom resource in Kubernetes, and if you create a
+custom resource in Kubernetes, it will also be available as a `HelmRelease`.
+
+We don't have any additional controllers that synchronize state between these resources.
+All requests to resources in our extension api-server are transparently proxied to `HelmRelease` and vice versa.
+This eliminates intermediate states and the need to write controllers and synchronization logic.
+
+## Implementation
+
+To implement the Aggregation API, you may consider two projects:
```

Comment:
Avoid implying these are the only two options.

Reply:
sure, fixed